v1/services/retention/service.go

```diff
@@ -0,0 +1,163 @@
+// Package retention provides the retention policy enforcement service.
+package retention // import "github.com/influxdata/influxdb/services/retention"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/influxdata/influxdb/v2/logger"
+	"github.com/influxdata/influxdb/v2/v1/services/meta"
+	"go.uber.org/zap"
+)
+
+// Service represents the retention policy enforcement service.
+type Service struct {
+	MetaClient interface {
+		Databases() []meta.DatabaseInfo
+		DeleteShardGroup(database, policy string, id uint64) error
+		PruneShardGroups() error
+	}
+	TSDBStore interface {
+		ShardIDs() []uint64
+		DeleteShard(shardID uint64) error
+	}
+
+	config Config
+	wg     sync.WaitGroup
+	done   chan struct{}
```

Comment:
Minor nit. This is probably just ported over from the previous implementation but I like using a cancelable background context instead of a `done` channel. It handles idempotency during closing and a lot of things already take a `context.Context`.

e.g.

```go
type Service struct {
    ctx context.Context
    cancel func()
}

func NewService() *Service {
    s := &Service{}
    s.ctx, s.cancel = context.WithCancel(context.Background())
    return s
}

Reply:
Good shout – I'll alter the `Open` method to take a `context.Context`